consensus/core/src/subscriber.rs

```diff
@@ -0,0 +1,244 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use std::{sync::Arc, time::Duration};
+
+use consensus_config::AuthorityIndex;
+use futures::StreamExt;
+use mysten_metrics::spawn_monitored_task;
+use parking_lot::{Mutex, RwLock};
+use tokio::{task::JoinHandle, time::sleep};
+use tracing::{debug, info};
+
+use crate::{
+    block::BlockAPI as _,
+    context::Context,
+    dag_state::DagState,
+    network::{NetworkClient, NetworkService},
+    Round,
+};
+
+/// Subscriber manages the block stream subscriptions to other peers, taking care of retrying
+/// if a subscription stream fails. Blocks returned from the peer are sent to authority service
+/// for processing.
+/// Currently individual subscription management is not exposed, but it could become
+/// useful in future.
+pub(crate) struct Subscriber<C: NetworkClient, S: NetworkService> {
+    context: Arc<Context>,
+    network_client: Arc<C>,
+    authority_service: Arc<S>,
+    dag_state: Arc<RwLock<DagState>>,
+    subscriptions: Arc<Mutex<Box<[Option<JoinHandle<()>>]>>>,
+}
+
+impl<C: NetworkClient, S: NetworkService> Subscriber<C, S> {
+    pub(crate) fn new(
+        context: Arc<Context>,
+        network_client: Arc<C>,
+        authority_service: Arc<S>,
+        dag_state: Arc<RwLock<DagState>>,
+    ) -> Self {
+        let subscriptions = (0..context.committee.size())
+            .map(|_| None)
+            .collect::<Vec<_>>();
```

Comment:
I might be missing something, but shouldn't exclude our selfs from the subscriptions? Or we init the vec for convenience so we do indexed access?